### DISCUSSION TITLE
Difference between IIncrementalGenerator RegisterImplementationSourceOutput vs RegisterSourceOutput

### DISCUSSION THREAD

**Discussion by TorreyGarland** [dotnet/roslyn#66885] [Q&A]
What is the difference between the two? 

The explanation provided by the documentation seemed vague.

Can you provide an example use case in the documentation?

**Comment by jasonmalinowski** [comment]
The difference is intended as an IDE optimization, although we haven't actually fully done all the work for it so the end-to-end isn't quite working yet. Generated code generally can be thought of as going in one of two buckets:

1. Code that defines new symbols, so that the developer's code can call into that. Consider something like the CsWin32 generator that generates methods to call.
2. Code that defines implementations of things, but there aren't new symbols. Consider a generator generating serialization/deserialization logic.

The IDE needs to run generators from step 1 to have accurate IntelliSense. It doesn't (usually) need to run the second one -- that's needed during the actual compile, but the IDE doesn't care about it most of the time. So "implementation source output" would only be run by the command line compiler for an actual build, or in rarer cases for the IDE, wheras the first method runs everywhere. The idea is if your generator only needs to generate implementation code, we can skip all of that in the IDE for performance reasons.

I'm hedging a bit about the IDE in that prior paragraph: we still would need to run the implementation source for edit and continue (since we're doing a "real" compile more or less in that case), and also if the user is trying to view generated code to set breakpoints, since you want to see the "real" code being ran in that case.

As of this writing we're still actually running both in the IDE though, but we're looking at fixing that properly soon.

**Reply by lsoft** [reply]
@jasonmalinowski let me ask about a related topic: what is the fate of `RegisterReferenceSourceOutput`? https://github.com/dotnet/roslyn/issues/57589

**Reply by jasonmalinowski** [reply]
@chsienki might have more to say; I think we generally concluded we should have it, and then just never found the time to actually do it.

**Comment by TorreyGarland** [comment]
How about a code example?

**Comment by ivandrofly** [comment]
> How about a code example?

See https://www.thinktecture.com/en/net/roslyn-source-generators-reduction-of-resource-consumption-in-ides/

**Comment by TorreyGarland** [comment]
```
At the time of writing, as I can see, the IDEs doesn’t treat RegisterImplementationSourceOutput any 
different than RegisterSourceOutput. But this should change in the future to be able to handle 
growing number of Roslyn Source Generators.
```

what is the point of using it then???

**Reply by AndersKehlet** [reply]
That if you use the appropriate one now, you'll get a perf bonus for free once they get around to fixing it?

**Reply by TorreyGarland** [reply]
When is it scheduled to be implemented?
